File: app/src/main/java/com/checkin/app/checkin/Shop/Private/ShopProfileViewModel.java
package com.checkin.app.checkin.Shop.Private;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.checkin.app.checkin.Shop.RestaurantModel;
import com.checkin.app.checkin.Shop.ShopRepository;
import com.checkin.app.checkin.data.BaseViewModel;
import com.checkin.app.checkin.data.Converters;
import com.checkin.app.checkin.data.network.RetrofitCallAsyncTask;
import com.checkin.app.checkin.data.notifications.MessageUtils;
import com.checkin.app.checkin.data.resource.Resource;
import com.checkin.app.checkin.misc.models.GenericDetailModel;
import com.checkin.app.checkin.utility.ProgressRequestBody;
import com.checkin.app.checkin.utility.SourceMappedLiveData;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.File;
import java.util.List;

/**
 * Created by Bhavik Patel on 24/08/2018.
 */

public class ShopProfileViewModel extends BaseViewModel {
    private ShopRepository mRepository;
    private long mShopPk;

    private SourceMappedLiveData<Resource<RestaurantModel>> mShopData = createNetworkLiveData();
    private MutableLiveData<Boolean> mCollectAspectData = new MutableLiveData<>();
    private MutableLiveData<Boolean> mCollectBasicData = new MutableLiveData<>();
    private MutableLiveData<JsonNode> mErrors = new MutableLiveData<>();

    public ShopProfileViewModel(@NonNull Application application) {
        super(application);
        mRepository = ShopRepository.Companion.getInstance(application);
        mCollectBasicData.setValue(false);
        mCollectAspectData.setValue(false);
    }

    public LiveData<Resource<RestaurantModel>> getShopData() {
        return mShopData;
    }

    public RestaurantModel updateAspectData(
            CharSequence[] cuisines, CharSequence[] categories, RestaurantModel.PAYMENT_MODE[] paymentModes,
            boolean hasNonVeg, boolean hasAlcohol, boolean hasHomeDelivery, List<String> extraData) {
        Resource<RestaurantModel> resource = mShopData.getValue();
        RestaurantModel shop;
        if (resource != null && resource.getStatus() == Resource.Status.SUCCESS && resource.getData() != null)
            shop = resource.getData();
        else
            shop = new RestaurantModel(mShopPk);
        shop.setCuisines(cuisines);
        shop.setCategories(categories);
        shop.setPaymentModes(paymentModes);
        shop.setHasAlcohol(hasAlcohol);
        shop.setHasHomeDelivery(hasHomeDelivery);
        shop.setHasNonveg(hasNonVeg);
        shop.setExtraData(extraData.toArray(new String[]{}));

        mShopData.setValue(Resource.Companion.success(shop));
        return shop;
    }

    public RestaurantModel updateBasicData(
            String name, String website, String tagLine,
            CharSequence[] nonWorkingDays, long openingTime, long closingTime) {
        Resource<RestaurantModel> resource = mShopData.getValue();
        RestaurantModel shop;
        if (resource != null && resource.getStatus() == Resource.Status.SUCCESS && resource.getData() != null)
            shop = resource.getData();
        else
            shop = new RestaurantModel(mShopPk);
        shop.setName(name);
        shop.setWebsite(website);
        shop.setTagline(tagLine);
        shop.setNonWorkingDays(nonWorkingDays);
        shop.setOpeningHour(openingTime);
        shop.setClosingHour(closingTime);

        mShopData.setValue(Resource.Companion.success(shop));
        return shop;
    }

    public LiveData<Boolean> shouldCollectAspectData() {
        return mCollectAspectData;
    }

    public LiveData<Boolean> shouldCollectBasicData() {
        return mCollectBasicData;
    }

    public void updateShopContact(String phoneToken, String email) {
        ObjectNode data = Converters.INSTANCE.getObjectMapper().createObjectNode();
        if (phoneToken != null)
            data.put("phone_token", phoneToken);
        if (email != null)
            data.put("email", email);
        getMData().addSource(mRepository.updateShopContact(mShopPk, data), getMData()::setValue);
    }

    public void updateShop(RestaurantModel shop) {
        getMData().addSource(mRepository.updateShopDetails(shop), getMData()::setValue);
    }

    public void collectData() {
        collectBasicData();
        collectAspectData();
    }

    public void collectBasicData() {
        mCollectBasicData.setValue(true);
    }

    public void collectAspectData() {
        mCollectAspectData.setValue(true);
    }

    public void showError(JsonNode data) {
        mErrors.setValue(data);
    }

    public LiveData<JsonNode> getErrors() {
        return mErrors;
    }

    public void fetchShopDetails(long shopPk) {
        mShopPk = shopPk;
        mShopData.addSource(mRepository.getShopModel(shopPk), mShopData::setValue);
    }

    public void fetchShopManage(long shopPk) {
        mShopPk = shopPk;
        mShopData.addSource(mRepository.getShopManageModel(shopPk), mShopData::setValue);
    }

    public long getShopPk() {
        return mShopPk;
    }

    public void setShopPk(long shopPk) {
        mShopPk = shopPk;
    }

    public void useShop(RestaurantModel restaurantModel) {
        mShopData.setValue(Resource.Companion.success(restaurantModel));
    }

    public void removeCoverImage(int index) {
        getMData().addSource(mRepository.deleteRestaurantCover(mShopPk, index), getMData()::setValue);
    }

    @Override
    public void updateResults() {
        fetchShopDetails(mShopPk);
    }

    public void updateCoverPic(File pictureFile, Context context, int index) {
        NotificationCompat.Builder builder = MessageUtils.createUploadNotification(context);
        MessageUtils.NotificationUpdate notificationUpdate = new MessageUtils.NotificationUpdate(context, builder);
        doUploadImage(pictureFile, notificationUpdate, index);
    }

    private void doUploadImage(File pictureFile, ProgressRequestBody.UploadCallbacks listener, int index) {
        if (index == -1)
            new RetrofitCallAsyncTask<GenericDetailModel>(listener).execute(mRepository.postRestaurantLogo(mShopPk, pictureFile, listener));
        else
            new RetrofitCallAsyncTask<GenericDetailModel>(listener).execute(mRepository.postRestaurantCover(mShopPk, index, pictureFile, listener));
    }
}

File: app/src/main/java/com/checkin/app/checkin/Shop/Private/ShopProfileViewModel.kt
package com.checkin.app.checkin.Shop.Private

import android.app.Application
import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.checkin.app.checkin.Shop.RestaurantModel
import com.checkin.app.checkin.Shop.ShopModel.PAYMENT_MODE
import com.checkin.app.checkin.Shop.ShopRepository
import com.checkin.app.checkin.data.BaseViewModel
import com.checkin.app.checkin.data.Converters.objectMapper
import com.checkin.app.checkin.data.network.RetrofitCallAsyncTask
import com.checkin.app.checkin.data.notifications.MessageUtils.NotificationUpdate
import com.checkin.app.checkin.data.notifications.MessageUtils.createUploadNotification
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.data.resource.Resource.Companion.success
import com.checkin.app.checkin.misc.models.GenericDetailModel
import com.checkin.app.checkin.utility.ProgressRequestBody.UploadCallbacks
import com.fasterxml.jackson.databind.JsonNode
import java.io.File

/**
 * Created by Bhavik Patel on 24/08/2018.
 */
class ShopProfileViewModel(application: Application) : BaseViewModel(application) {
    private val mRepository: ShopRepository = ShopRepository.getInstance(application)

    private val mShopData = createNetworkLiveData<RestaurantModel>()
    private val mCollectAspectData = MutableLiveData<Boolean>(false)
    private val mCollectBasicData = MutableLiveData<Boolean>(false)
    private val mErrors = MutableLiveData<JsonNode>()

    var shopPk: Long = 0
    val shopData: LiveData<Resource<RestaurantModel>> = mShopData
    val errors: LiveData<JsonNode> = mErrors

    fun updateAspectData(
            cuisines: Array<CharSequence>, categories: Array<CharSequence>, paymentModes: Array<PAYMENT_MODE>,
            hasNonVeg: Boolean, hasAlcohol: Boolean, hasHomeDelivery: Boolean, extraData: List<String>): RestaurantModel {
        val shop = mShopData.value?.data ?: RestaurantModel(shopPk)
        shop.setCuisines(*cuisines)
        shop.setCategories(*categories)
        shop.setPaymentModes(*paymentModes)
        shop.setHasAlcohol(hasAlcohol)
        shop.setHasHomeDelivery(hasHomeDelivery)
        shop.setHasNonveg(hasNonVeg)
        shop.setExtraData(*extraData.toTypedArray())
        mShopData.value = success(shop)
        return shop
    }

    fun updateBasicData(
            name: String, website: String, tagLine: String,
            nonWorkingDays: Array<CharSequence>, openingTime: Long, closingTime: Long): RestaurantModel {
        val shop = mShopData.value?.data ?: RestaurantModel(shopPk)
        shop.name = name
        shop.website = website
        shop.tagline = tagLine
        shop.nonWorkingDays = nonWorkingDays
        shop.openingHour = openingTime
        shop.closingHour = if (closingTime < openingTime) closingTime + 24 * 60.toLong() else closingTime
        mShopData.value = success(shop)
        return shop
    }

    fun shouldCollectAspectData(): LiveData<Boolean> {
        return mCollectAspectData
    }

    fun shouldCollectBasicData(): LiveData<Boolean> {
        return mCollectBasicData
    }

    fun updateShopContact(phoneToken: String?, email: String?) {
        val data = objectMapper.createObjectNode().apply {
            if (phoneToken != null) put("phone_token", phoneToken)
            if (email != null) put("email", email)
        }
        mData.addSource(mRepository.updateShopContact(shopPk, data), mData::setValue)
    }

    fun updateShop(shop: RestaurantModel) = mData.addSource(mRepository.updateShopDetails(shop), mData::setValue)

    fun collectData() {
        collectBasicData()
        collectAspectData()
    }

    fun collectBasicData() {
        mCollectBasicData.value = true
    }

    fun collectAspectData() {
        mCollectAspectData.value = true
    }

    fun showError(data: JsonNode) {
        mErrors.value = data
    }

    fun fetchShopDetails(shopPk: Long) {
        this.shopPk = shopPk
        mShopData.addSource(mRepository.getShopModel(shopPk)) { value: Resource<RestaurantModel> -> mShopData.setValue(value) }
    }

    fun fetchShopManage(shopPk: Long) {
        this.shopPk = shopPk
        mShopData.addSource(mRepository.getShopManageModel(shopPk), mShopData::setValue)
    }

    fun useShop(restaurantModel: RestaurantModel?) {
        mShopData.value = success(restaurantModel)
    }

    fun removeCoverImage(index: Int) {
        mData.addSource(mRepository.deleteRestaurantCover(shopPk, index), mData::setValue)
    }

    override fun updateResults() {
        fetchShopDetails(shopPk)
    }

    fun updateCoverPic(pictureFile: File, context: Context?, index: Int) {
        val builder = createUploadNotification(context!!)
        doUploadImage(pictureFile, NotificationUpdate(context, builder), index)
    }

    private fun doUploadImage(pictureFile: File, listener: UploadCallbacks<GenericDetailModel>, index: Int) {
        if (index == -1) RetrofitCallAsyncTask(listener).execute(mRepository.postRestaurantLogo(shopPk, pictureFile, listener))
        else RetrofitCallAsyncTask(listener).execute(mRepository.postRestaurantCover(shopPk, index, pictureFile, listener))
    }
}